NEXT-12694 - Add custom rules (#131)
shopware/docs: guides/plugins/plugins/framework/rule/add-custom-rules.md

# Add custom rules

## Overview

> In this guide you'll learn how to create rules in Shopware. Rules are used by the Rule builder.

Reviewer: ```suggestion
In this guide you'll learn how to create rules in Shopware. Rules are used by the rule builder.
```

        }

        // Shop administrator wants the rule to match if there's currently NOT a lunar eclipse.
        return !$isCurrentlyLunarEclipse;
    }

    public function getConstraints(): array
    {
        return [
            'isLunarEclipse' => [ new Type('bool') ]
        ];
    }
}
```
{% endcode %}

As you can see, several methods are already implemented:

- `__constructor`: This only defines the default expected value. This is overwritten at runtime with the actual value, that the shop owner set in the administration.
- `getName`: Returns a unique technical name for your rule.
- `match`: This checks whether the rule applies. Accordingly, a boolean is returned whether the rule applies or not.
- `getConstraints`: This method returns an array of the possible fields and its types. You could also return the `NotBlank` class here, to require this field.

After we've created our rule class, we have to register it in our `services.xml` and tag it as `shopware.rule.definition`.

### Active rules

You can access all active rules by using the `getRuleIds` method of the context.

```php
$context->getRuleIds();
```

### Showing rule in the administration

> Now we want to implement our new rule in the administration that we can manage it. To achieve this, we have to call the `addCondition` method of the [RuleConditionService](https://github.com/shopware/platform/blob/v6.3.4.1/src/Administration/Resources/app/administration/src/app/service/rule-condition.service.js), by decorating this service.

Reviewer: ```suggestion
Now we want to implement our new rule in the administration so that we can manage it. To achieve this, we have to call the `addCondition` method of the [RuleConditionService](https://github.com/shopware/platform/blob/v6.3.4.1/src/Administration/Resources/app/administration/src/app/service/rule-condition.service.js), by decorating this service.
```

> Now we want to implement our new rule in the administration so that we can manage it. To achieve this, we have to call the `addCondition` method of the [RuleConditionService](https://github.com/shopware/platform/blob/v6.3.4.1/src/Administration/Resources/app/administration/src/app/service/rule-condition.service.js), by decorating this service.

Reviewer 2: Maybe link the "Decorate a service" guide. Iirc it's called "customize-service" or something like that

> To do this, we create a new directory called `<plugin root>/src/Resources/app/administration/src/decorator`. In this directory we create a new file called `rule-condition-service-decoration.js`.

Reviewer: You begin your sentences with:
To achive this [...]
To do this [...]

Change this here to something else, or omit it at all:
```suggestion
Create a new directory called `<plugin root>/src/Resources/app/administration/src/decorator`. In this directory we create a new file called `rule-condition-service-decoration.js`.
```

{% code title="<plugin root>/src/Resources/app/administration/src/decorator/rule-condition-service-decoration.js" %}
```javascript
import '../core/component/swag-lunar-eclipse';

Shopware.Application.addServiceProviderDecorator('ruleConditionDataProviderService', (ruleConditionService) => {
    ruleConditionService.addCondition('lunar_eclipse', {
        component: 'swag-lunar-eclipse',
        label: 'Is lunar eclipse today',
        scopes: ['global']
    });

    return ruleConditionService;
});
```
{% endcode %}

> As you can see, this is decorating the `RuleConditionService` by using its name `ruleConditionDataProviderService`, which is defined here. 

Reviewer 2: "which is defined here"... Here? :-P

> The decoration adds a new condition called `lunar_eclipse` then. Make sure to match the name we've used in the `getName` method in PHP. 

Reviewer: ```suggestion
The decoration adds a new condition called `lunar_eclipse`. Make sure to match the name we've used in the `getName` method in PHP. 
```

> It comes with a custom component `swag-lunar-eclipse`, which we will create in the next step. 

Reviewer: ```suggestion
Next, we define the component, in our case `swag-lunar-eclipse`, which is responsible for rendering the rule inside the administration. We will create this component in the next step. 
```

We will create this component in the next step. Furthermore we defined a label, which will be displayed in the rule builder selection.
The last option is the scope, which in our case is `global`, as we have not specified a specific one in our core class.

### Custom rule component

Since we've registered our rule to the administration now, we're still lacking the actual component `swag-lunar-eclipse`. 
As previously mentioned, we've already defined a path for it in our service decoration.

> So create the following directory: `<plugin root>/src/Resources/app/administration/src/core/component/swag-lunar-eclipse`.

Reviewer 2: Link to "create component" guide

> Here's an example of what this component could look like, with an explanation coming afterwards:

Reviewer: ```suggestion
Here's an example of what this component could look like:
```

                }

                return this.condition.value.isLunarEclipse;
            },
            set(isLunarEclipse) {
                this.ensureValueExist();
                this.condition.value = { ...this.condition.value, isLunarEclipse };
            }
        }
    }
});
```
{% endcode %}

As you can see, our `swag-lunar-eclipse` has to extend from the `sw-condition-base` component and has to bring a custom template, which will be explained in the next step.
Let's have a look at each property and method. The first computed property is `selectValues`, which returns an array containing the values "true" and "false". Those will be used in the template later on, as they will be the selectable options for the shop administrator. Do not get confused by the call this.$tc('global.sw-condition.condition.yes'), it's just loading a translation by its name, in this case "Yes" and "No". Note: When dealing with boolean values, make sure to always return strings here!

The second and last computed property is `isLunarEclipse`, which uses a getter and setter to define the value of the condition.

### Custom rule administration template

> The last step is, creating a template for our condition. We will create a new file called `swag-lunar-eclipse.html.twig` in the directory `<plugin root>/src/Resources/app/administration/src/core/component/swag-lunar-eclipse/swag-lunar-eclipse.html.twig`.

Reviewer: You could simply omit the path here, as it is the same ? 

```suggestion
The last step is creating a template for our condition. We will create a new file called `swag-lunar-eclipse.html.twig` in the same directory as the component.
```

In our template, we have to overwrite the block `sw_condition_value_content`. In this example we define a `sw-single-select` in this block.

{% code title="<plugin root>/src/Resources/app/administration/src/core/component/swag-lunar-eclipse/swag-lunar-eclipse.html.twig" %}
```twig
{% block sw_condition_value_content %}
    <sw-single-select name="lunar-eclipse"
                      id="lunar-eclipse"
                      size="medium"
                      :options="selectValues"
                      v-model="isLunarEclipse"
                      class="field--main">
    </sw-single-select>
{% endblock %}
```
{% endcode %}

As you can see, our `sw-single-select` uses the previously created computed property `selectValues` as the `options` prop, and the value is saved into the variable `isLunarEclipse`.

> That's it, now your rule is fully integrated.

Reviewer: ```suggestion
That's it, your rule is now fully integrated.
```

- `__constructor`: This only defines the default expected value. This is overwritten at runtime with the actual value, that the shop owner set in the administration.
- `getName`: Returns a unique technical name for your rule.
- `match`: This checks whether the rule applies. Accordingly, a boolean is returned whether the rule applies or not.
- `getConstraints`: This method returns an array of the possible fields and its types. You could also return the `NotBlank` class here, to require this field.

After we've created our rule class, we have to register it in our `services.xml` and tag it as `shopware.rule.definition`.

### Active rules

You can access all active rules by using the `getRuleIds` method of the context.

```php
$context->getRuleIds();
```

### Showing rule in the administration

Now we want to implement our new rule in the administration so that we can manage it. To achieve this, we have to call the `addCondition` method of the [RuleConditionService](https://github.com/shopware/platform/blob/v6.3.4.1/src/Administration/Resources/app/administration/src/app/service/rule-condition.service.js), by decorating this service.

> The decoration of services in the administration will be covered in our [Adding services](../../administration/add-custom-service.md) guide.

Reviewer 2: ```suggestion
The decoration of services in the administration will be covered in our [Adding services](../../administration/add-custom-service.md#Decorating a service) guide.
```
